### ISSUE TITLE
Universal save/load function

### ISSUE THREAD

**Issue by railsmechanic** [fredwu/simple_bayes#3]
In order to save the bayes filter e.g. in a database/key-value store I think an universal save/load function would be great.

My suggestion is to adapt the storage behavior so that the save function returns a `{:ok, pid, data}` tuple. In case of the filesystem storage, which doesn't need to return data, the returned tuple is `{:ok, pid, nil}`. 

The load function than needs to accepts the encoded data.

What do you think?

**Comment by fredwu** [comment]
Would you be able to clarify the rational behind this - what's the advantage of returning `data`?

**Comment by railsmechanic** [comment]
The filesystem storage is implemented to write the struct directly to a file and return the agent pid.
My suggestion is to return the encoded struct directly, to be able to save it e.g. in a database/key-value store. **(Sample see below)**

``` elixir
defmodule SimpleBayes.Storage.Universal do
  @behaviour SimpleBayes.Storage.Behaviour

  alias SimpleBayes.Data

  def init(struct, _opts) do
    {:ok, pid} = Agent.start_link(fn -> struct end)

    pid
  end

  def save(pid, struct) do
    {:ok, pid, Data.encode(struct)}
  end

  def load(encoded_data, opts) do
    struct = Data.decode(encoded_data)
    init(struct, opts)
  end
end
```

I hope this example explains my intention. Are their any other options to save the struct data?

**Comment by fredwu** [comment]
I see, if the intention is to get the encoded data, I can return the struct as you suggested in `save`, and one could simply use the memory storage for it (instead of a different storage backend). I am however not sure about passing the struct to `load` - as the interface for all storage backends should be the same. What do you think?

**Comment by fredwu** [comment]
My brain isn't working... of course there needs to be a way to load the struct. I'll make `load` take an optional struct. :)

**Comment by railsmechanic** [comment]
Your're answering faster than I can type 😄 

I think taking an optional struct is the way to go. 👍

**Comment by fredwu** [comment]
Just released v0.9.0 with this feature, please check it out. :)
